instrumentation/startup/src/main/java/io/opentelemetry/android/instrumentation/startup/SdkInitializationEvents.kt

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.android.instrumentation.startup
+
+import io.opentelemetry.android.common.RumConstants
+import io.opentelemetry.api.common.AttributeKey
+import io.opentelemetry.api.common.Attributes
+import io.opentelemetry.api.incubator.logs.AnyValue
+import io.opentelemetry.sdk.OpenTelemetrySdk
+import io.opentelemetry.sdk.logs.internal.SdkEventLoggerProvider
+import io.opentelemetry.sdk.trace.export.SpanExporter
+import java.time.Instant
+import java.util.function.Consumer
+import java.util.function.Supplier
+
+class SdkInitializationEvents(private val clock: Supplier<Instant> = Supplier { Instant.now() }) : InitializationEvents {
+    private val events: MutableList<Event> = ArrayList()
+
+    override fun sdkInitializationStarted() {
+        addEvent(RumConstants.Events.INIT_EVENT_STARTED)
+    }
+
+    override fun recordConfiguration(config: Map<String, String>) {
+        val map: MutableMap<String, AnyValue<*>> = HashMap()
+        config.entries.forEach(
+            Consumer { e: Map.Entry<String, String> ->
+                map[e.key] = AnyValue.of(e.value)
+            },
+        )
+        val body = AnyValue.of(map)
+        addEvent(RumConstants.Events.INIT_EVENT_CONFIG, body)
+    }
+
+    override fun currentNetworkProviderInitialized() {
+        addEvent(RumConstants.Events.INIT_EVENT_NET_PROVIDER)
+    }
+
+    override fun networkMonitorInitialized() {
+        addEvent(RumConstants.Events.INIT_EVENT_NET_MONITOR)
+    }
+
+    override fun anrMonitorInitialized() {
+        addEvent(RumConstants.Events.INIT_EVENT_ANR_MONITOR)
+    }
+
+    override fun slowRenderingDetectorInitialized() {
+        addEvent(RumConstants.Events.INIT_EVENT_JANK_MONITOR)
+    }
+
+    override fun crashReportingInitialized() {
+        addEvent(RumConstants.Events.INIT_EVENT_CRASH_REPORTER)
+    }
+
+    override fun spanExporterInitialized(spanExporter: SpanExporter) {
+        val attributes =
+            Attributes.of(AttributeKey.stringKey("span.exporter"), spanExporter.toString())
+        addEvent(RumConstants.Events.INIT_EVENT_SPAN_EXPORTER, attributes)
+    }
+
+    fun finish(sdk: OpenTelemetrySdk) {
+        val loggerProvider = sdk.sdkLoggerProvider
+        val eventLogger =
+            SdkEventLoggerProvider.create(loggerProvider).get("otel.initialization.events")
+        events.forEach { event: Event ->
+            val eventBuilder =
+                eventLogger.builder(event.name)
+                    .setTimestamp(event.timestamp)
+                    .setAttributes(event.attributes)
+            if (event.body != null) {
+                // TODO: Config is technically correct because config is the only startup event
+                // with a body, but this is ultimately clunky/fragile.
+                eventBuilder.put("config", event.body)
+            }
+            eventBuilder.emit()
```

Comment:
I think I misunderstood the goal of the `InitializationEvents` at first as I thought of it as a set of callbacks for ourselves and our users to keep track of the state of the RUM initialization in code, though I wasn't expecting this feature to actually generate telemetry, which brings a couple of questions. Do we need to generate telemetry for this use-case (or is it enough to provide the callbacks and then our users can do anything they'd wish there, for example sending their own events if needed)? Can users opt out of it? Do we need to define the events in the semantic conventions first? Since this feature is automatically generating telemetry, shouldn't it be an `AndroidInstrumentation` implementation?

Reply:
> Do we need to generate telemetry for this use-case (or is it enough to provide the callbacks and then our users can do anything they'd wish there, for example sending their own events if needed)?

Generating events that track milestones in the agent startup and configuration is the main goal with this. These events have been extremely helpful when looking at startup times and understanding how long the agent itself takes to reach certain states in the initialization. I don't see any benefit in exposing these hooks to user/application code. What would a user do at any of these times? They can't send their own events, because the sdk has not been initialized.

> Can users opt out of it?

Yes, there is `config.disableSdkInitializationEvents()`.

> Do we need to define the events in the semantic conventions first?

I don't think it has to happen first, but yes, it would be correct to have defined semantic conventions for these.

> Since this feature is automatically generating telemetry, shouldn't it be an AndroidInstrumentation implementation

Almost certainly yes, but that didn't exist when I wrote it! :) Moving so fast over here....heh. We can pick that up as a follow-on item I hope.

Reply:
> They can't send their own events, because the sdk has not been initialized.

We technically can't either :) though there's always a way... In any case, I get the idea now, it's certainly helpful to get this info for debugging the Android agent 👍 

I'm generally hesitant about producing telemetry out of the box because it assumes that we know what every user of this lib will want to emit, which sounds a bit opinionated, so it's nice that we provide a way to opt-out.

> I don't think it has to happen first, but yes, it would be correct to have defined semantic conventions for these.

This use case seems to be pretty specific to this agent, so I'm fine with moving forward without adding these events to the semconv first, however, I'd like to have some sort of litmus test to know when it's fine to send telemetry without conventions and when it's not, because we're going to need to be prepared when someone asks about it (unless it's a common practice around OTel impls to send telemetry that's not yet defined in the conventions, in which case then we probably shouldn't stress too much about it).

> Almost certainly yes, but that didn't exist when I wrote it! :) Moving so fast over here....heh. We can pick that up as a follow-on item I hope.

Sounds good! It seems like you need to have this feature available before the instrumentation API work is done, so I think it's fine to move forward as is for now. I'm planning to adapt all the existing automatically generated telemetry to the new API anyway, so I think I can pick this one up too.